### PR TITLE
docs(ssr): add experimental app directory guide

### PR DESCRIPTION
**Pull request by efilion**
Adding Next.js 13 app directory docs discussed in https://github.com/TanStack/query/discussions/4558.

The initialData approach was first published by @lukemorales. This approach appears to be subject to the same caveats as with using initialData in the standard `pages` directory. Having to thread `initialData` through to a component deep down in a component tree is aggravated by now having to fetch data in an ancestor Server Component.

Given those drawbacks, I've also documented an approach that rehydrates the Query Cache. This approach was largely informed by [Using Other Frameworks or Custom SSR Frameworks](https://tanstack.com/query/v4/docs/guides/ssr#using-other-frameworks-or-custom-ssr-frameworks). There are two notable deviations from that guide:
- Not calling `queryClient.clear()`. As per @TkDodo, this is no longer necessary since v4.
- Not populating window.__REACT_QUERY_STATE__ with dehydratedState. Next.js is handling hydration without any need for this.

### CODE REVIEW

File: docs/guides/ssr.md
Comment (TkDodo) [inline docs/guides/ssr.md:200]: I think we should be able to decouple `QueryClientProvider` and `Hydrate`. The Provider needs to go at the top, once. Nothing changes here really.

`Hydrate` can appear multiple times in the app tree. It just takes data and puts it into the cache if it's newer, as opposed to `initialData`, which will only write data once to the cache if there is no data yet.

**Review by TkDodo — COMMENTED** [review]
Thanks for this. I would like to structure it a bit differently:

let's keep everything as it is for the nextJs (12) docs and start a new section at the bottom of this page called `Experimental app directory in Next 13` or something. There, we can put the `initialData` and `Hydrate` approaches, with another notice that this is experimental and best practices might change as we figure things out :)

File: docs/guides/ssr.md
Comment (efilion) [reply docs/guides/ssr.md:200]: Decoupling it is a really good point. In the current implementation, you might end up prefetching data at the top level for a leaf component that is conditionally rendered. Effectively, you could end up prefetching data that isn't needed in the end. Fetching data closer to the component that needs it could also allow you to parameterize that fetch with arguments that wouldn't be available at the top level.

I'll play around with these ideas a bit and see what kind of new implementation I can get working. Thinking something along the lines of a higher-order component that does the fetching and wraps its argument in a `Hydrate`.

**Comment by efilion** [comment]
> let's keep everything as it is for the nextJs (12) docs and start a new section at the bottom of this page called `Experimental app directory in Next 13` or something. There, we can put the `initialData` and `Hydrate` approaches, with another notice that this is experimental and best practices might change as we figure things out :)

Agreed. My instinct was to have them together because they share things in common, and I didn't want to be too repetitive, but I'm not writing an essay here and the goal isn't to compare and contrast. When people read this, they'll just choose to read one section, so being repetitive isn't as bad. Better not to disrupt the flow of documentation that people are already used to or to introduce experimental stuff before fully addressing the more standard approach.

File: docs/guides/ssr.md
Comment (KATT) [inline docs/guides/ssr.md:222]: Seems really backwards that you gotta `await queryClient.prefetchQuery(['posts'], getPosts)` in order for the state to be hydrated... 

I would expect some magic to hydrate anything that has been called with `useQuery(..., { suspense: true })` automatically

Comment (KATT) [reply docs/guides/ssr.md:222]: Is there a missing API in React for this at the moment? No way of pushing state from a server-render React-component to a client-side one without this stuff?

File: docs/guides/ssr.md
Comment (KATT) [inline docs/guides/ssr.md:183]: I'd add a big `:::caution` under here that this is experimental and is subject to change awaiting _{{some link to a React-issue or RFC}}_

File: docs/guides/ssr.md
Comment (KATT) [inline docs/guides/ssr.md:210]: I'd rephrase this somehow

```suggestion
- Create a new QueryClient instance **within the body of the React component**. This ensures that data is not shared between users and requests.
```

File: docs/guides/ssr.md
Comment (efilion) [reply docs/guides/ssr.md:222]: Server-side suspense is a whole other wrinkle I hadn't considered. Dropping https://github.com/reactwg/react-18/discussions/37 here for reference because I'll likely want to circle back on this.

Comment (TkDodo) [reply docs/guides/ssr.md:222]: I agree that this could be handled better by TanStack/query, but for now, it is what it is :). This is exactly why I'd like the whole section to be marked as `experimental` - because the recommendations might change.

also looping in @Ephem :)

File: docs/guides/ssr.md
Comment (Ephem) [inline docs/guides/ssr.md:27]: Love this section! It's been sorely missing, glad you included it, and well written. :smile:

File: docs/guides/ssr.md
Comment (Ephem) [reply docs/guides/ssr.md:200]: I'd say keep it simple for now, the API is probably going to change in the next RQ major anyway. If you wrap `QueryClientProvider` higher up in the tree, you can just do:

```jsx
'use client';
import { Hydrate } from '@tanstack/react-query'

export default function HydrateOnClient(dehydratedState) {
  return <Hydrate state={dehydratedState}>{children}</Hydrate>;
}
```

and use that directly in the Server Components. Don't think there is need for a HoC or something more complex here (at least yet).

File: docs/guides/ssr.md
Comment (Ephem) [reply docs/guides/ssr.md:222]: I mean, the recommendation should always be to prefetch things as early as possible, both to get a headstart and to avoid waterfalls where you suspend in A to be able to render B which then suspend again etc. We should definitely investigate supporting the other case as well though (if it's even going to be possible), but that's for the future.

One caveat is that currently you _have_ to await the data in the RSC, which blocks streaming what you can early. What we really want is to _start_ prefetching in the RSC but not await it, instead suspending in the component that needs this data. That will require both that React implements de/hydration of Promises and a bunch of changes to RQ though, so I think this PR (with the proposed changes elsewhere) documents the best current experimental approach.

**Review by Ephem — COMMENTED** [review]
Good work on this so far! In a longer time perspective I think we have a lot to figure out about how to document and teach this for different cases, frameworks etc, but don't think there is much use in doing that before we figure out what the "next generation" of hydration etc is going to look like, this looks like a good stopgap for now for people that want to experiment. :smile:

File: docs/guides/ssr.md
Comment (Ephem) [reply docs/guides/ssr.md:222]: Oh, and like we talked about on Discord @KATT, I think it's interesting here to think about how to support "normal" streaming rendering without RSCs as well. A lot of frameworks might implement streaming rendering without RSCs, so that's also something we want to support.

File: docs/guides/ssr.md
Comment (efilion) [reply docs/guides/ssr.md:200]: https://codesandbox.io/p/github/efilion/next13-react-query/decoupled-hydrate?file=%2Fapp%2Fpage.tsx

The HoC isn't strictly necessary, but it helps hide some of the complexity and keep things DRY.

To take advantage of the deduplication provided by QueryClient, the instance defined at the page level will need to be threaded down through to all components looking to prefetch data. That nullifies one of the advantages of the Hydration approach over the InitialData approach. It would be nice to have something akin to an IoC container that could provide a QueryClient as a request-scoped singleton, but I didn't see much in the way of options to accomplish that at the moment. InversifyJS and TSyringe don't fit the bill here. Likely a difficulty with the fact that being single-threaded, Node.js doesn't provide each request its own thread. Nest.js has a solution, so maybe Next.js could implement this itself down the line. Alternatively, can look at the Reader Monad pattern from functional programming to handle dependency injection.

Comment (Ephem) [reply docs/guides/ssr.md:200]: Yeah, that's a tricky one, clever solution!

I worry about putting that much complexity in the docs though, especially for an experimental approach, might fit better as an example to link to? There might be another way as well, not saying it's better but I thought I'd throw it out there.

So behind the scenes we have RSCs, client (server rendered) and client (browser rendered). For both of the client parts, it's important to have the same `queryClient` throughout the whole rendering which is why we want to hoist it to the top, but for the RSCs, your initial approach is actually fine where you create a new `queryClient` for each RSC (fine as in works, not as in pretty, DRY, fast etc). The reason for this is that since that data gets de/rehydrated anyway, we only need the `queryClient` as a way to be able to call `prefetchQuery` and get the right format back when we call `dehydrate`. The fetches themselves will be deduped behind the scenes anyway, so even if two RSCs call the same endpoint, it at least wont double fetch.

So for RSCs specifically, it's fine to duplicate the `queryClient` instead of passing it around. Ugly and not something we want long term, but might be easier and more concise to document, although possibly more confusing?

File: docs/guides/ssr.md
Comment (Ephem) [reply docs/guides/ssr.md:222]: Btw, for now we should definitely document that `useQuery(..., { suspense: true })` wont work on the server unless you have already prefetched that data.

File: docs/guides/ssr.md
Comment (efilion) [reply docs/guides/ssr.md:200]: > > but it helps hide some of the complexity and keep things DRY.
> 
> I guess another way of stating the above is that I'm not sure this is something we should aim to do at this point, I kind of like having the full extent of the experimental and raw nature of this fully visible, it becomes less of an "officially endorsed pattern" that way which is something that might bite us later (even with caveats, some devs will just copy-paste it). sweat_smile
> 
> Maybe I worry too much though.

I see what you mean. My HoC approach could be a better match for a blog post.

Comment (efilion) [reply docs/guides/ssr.md:200]: https://codesandbox.io/p/github/efilion/next13-react-query/hydrated-content?file=%2Fapp%2FhydratedContent.tsx

Mainly looking to improve on line 12 inside `hydratedContent.tsx`. Not setting anything at all here might be the best approach for the moment. Might need to make a change to `dehydrate` to pick and choose which queries to dehydrate based on queryKey. Open to suggestions.

File: docs/guides/ssr.md
Comment (efilion) [reply docs/guides/ssr.md:222]: If you've already prefetched your data, is there any utility to set a client-side suspense boundary? If you don't prefetch, you can set suspense boundaries inside a client-side component and use `useQuery(..., { suspense: true })`. Data will start fetching after the page loads. If you're prefetching, you can set a suspense boundary in a server-side component, and there's no need for `{ suspense: true }` inside `useQuery`. Data will start fetching as soon as possible, the server will send a response without waiting for it, and then stream the rendered component to be integrated into the page once the fetching completes.

**Comment by efilion** [comment]
Line 218: Wondering if it would be a good idea to mention that eventually 'use client' will be added to `react-query` modules or that this is still being evaluated.

**Comment by efilion** [comment]
Line 248: Looking only to dehydrate the queries that were prefetched in the Server Component. Could export `matchQuery` for `query-core`, but will need to make the types match, and time complexity would be less than ideal. Alternatively, could expand the dehydrate API to accept a list of queryKeys that can be looked up by queryHash.

**Comment by Ephem** [comment]
I haven't had time yet to look at the latest changes here (soon!), but I thought I'd note that I've written a braindump discussion ["React Streaming SSR/Server Components"](https://github.com/TanStack/query/discussions/4623) that relates to this PR.

**Comment by efilion** [comment]
> I haven't had time yet to look at the latest changes here (soon!), but I thought I'd note that I've written a braindump discussion ["React Streaming SSR/Server Components"](https://github.com/TanStack/query/discussions/4623) that relates to this PR.

Thanks for writing this up! Tons of really useful background/context. Will take a bit for me to digest all of it and get up to speed.

Already, I'm thinking I should just drop the query cache filtering on line 248, and wait to see what new API will be settled on. Maybe I make a note that the current approach serializes and sends more duplicate data to the client than necessary, but that this will be addressed in a later update.

Also, I think I still need to think more thoroughly about what will happen with my current approach if someone opts in to concurrent features... especially Suspense. I can see that the query cache could look different between the time a component renders on the server and the time it hydrates on the client. Two components share a query that wasn't prefetched on the server, one component hydrates and fetches the query before the second can hydrate. Even _refetching_ of a query that was initially prefetched on the server could be problematic.

**Comment by Ephem** [comment]
> sends more duplicate data to the client than necessary

Yeah, this can be a tradeoff in some cases, but I think compression (gzip/brotli) might take care of a lot of that since the data would be identical, and I'm not sure if React itself might even try to do something smart here when props are identical? Still something to keep an eye on though.

> I can see that the query cache could look different between the time a component renders on the server and the time it hydrates on the client

When we fix `useSyncExternalStore should have a correct/frozen getServerSnapshot (minor)`, that will take care of this. That makes it so during hydration, whenever that happens, all RQ hooks etc reads the _original_ data that was passed down from the server, even if it has later changed. :smiley:

File: docs/guides/ssr.md
Comment (Necmttn) [reply docs/guides/ssr.md:165]: Good catch 🕺

Comment (atilafassina) [reply docs/guides/ssr.md:165]: wow, big TIL, @efilion. Thanks for that!

It still looks to me that even when that change lands, it will be a workaround based on specific use-cases. If it's not necessary, I'd still suggest to omit it.

You make a good point about `create-next-app` though. I'm not aware of **AppDir** as part of the CLI template though. I'd agree 100% in keeping it consistent with the default template if there's already one.

File: docs/guides/ssr.md
Comment (TkDodo) [reply docs/guides/ssr.md:218]: yes, we have plans to do this. @Ephem this is something we could even do for v4, right?

File: docs/guides/ssr.md
Comment (TkDodo) [reply docs/guides/ssr.md:248]: agree, we should probably just use the default settings here. There likely are no mutations, so setting `dehydrateMutations` to `false` doesn't accomplish much

**Comment by HamAndRock** [comment]
Hey, trying to use the Hydrate approach but getting this error, when adding the dehydrate() step to server component, runninng "@tanstack/react-query": "^4.22.0".

image

Is this supposed to happen since this is still a draft ?

**Comment by Ephem** [comment]
@HamAndRock That's weird. Some questions:

* Does this happen even if you only call `dehydrate`, but not the `Hydrate` component?
* If yes, where are you importing `dehydrate` from? It lives in the `@tanstack/query-core` package, so just importing and using that should absolutely not give you that error (but it's also re-exported in `@tanstack/react-query` which is why I'm asking.
* If it only happens with the `Hydrate` component, did you remember to wrap it in a `HydrateOnClient` component with `'use client'` on top?

**Comment by HamAndRock** [comment]
Hey, thank you for answer, the issue was in the import of dehydrate like you mentioned I used autocomplete to import it and it picked the react-query version instead of query-core.

I am sorry, this is my fault for wasting your time. I suggest maybe adding some kind of warning to the docs about these two "identical" imports, so it's easier to find the issue faster.

Thank you again and sorry for me being not able to copy code :D

**Comment by TkDodo** [comment]
@Ephem @efilion what would it take for us to get this from a draft PR to something that we can merge?

**Comment by efilion** [comment]
> Hey, thank you for answer, the issue was in the import of dehydrate like you mentioned I used autocomplete to import it and it picked the react-query version instead of query-core.
> 
> I am sorry, this is my fault for wasting your time. I suggest maybe adding some kind of warning to the docs about these two "identical" imports, so it's easier to find the issue faster.
> 
> Thank you again and sorry for me being not able to copy code :D

Not wasting our time at all. Really valuable feedback into improving DX of the docs. I think adding that kind of warning is a good idea. As you mentioned, most IDEs will autocomplete that line to use the wrong module, and it's very easy for our eyes to glaze over such a small detail.

**Comment by efilion** [comment]
> @Ephem @efilion what would it take for us to get this from a draft PR to something that we can merge?

1. - [ ] Drop dehydrate options
- As @Ephem mentioned, compression mitigates some of the problems with sending duplicate data to the client. I don't think this is a detail that needs to be added to the docs, but this issue may be a worthwhile consideration for future versions of the library. 
- I will make this change shortly.

2. - [ ] Make layout.jsx more consistent with create-next-app template
- I will make this change shortly.

3. - [ ] Mention that v4 targets a fix for having to re-export `<Hydrate>` simply to add `use client`
- Waiting for confirmation on this. Alternatively, we can opt not mention anything in the docs at the present time.

4. - [ ] Add a note on importing from react-query vs query-core
- I will make this change shortly.

5. - [ ] Further investigation/testing into the behaviour of Suspense and other concurrent features
- I think this can be made optional for now by stressing in the Suspense section that support for such is still experimental. Maybe add a prompt to create an issue if unexpected behaviour is encountered.

I think we have a PR that adds a lot of utility as is, so I would suggest merging it after todos 1-4 are completed. Any refinements based on todo 5 could be addressed in a later PR.

**Comment by TkDodo** [comment]
> Waiting for confirmation on this. Alternatively, we can opt not mention anything in the docs at the present time.


We have a PR that adds `useClient` everywhere where we use state, context or effects:
- https://github.com/TanStack/query/pull/4738

I think the PR is good to go, but if you want to give it a review, please do. Would this solve the problem? 

> I think this can be made optional for now by stressing in the Suspense section that support for such is still experimental.

👍

**Comment by Ephem** [comment]
I think we can land https://github.com/TanStack/query/pull/4738 first, I think it needs some changes but should be pretty quick?

So if you want to modify this to remove the `HydrateOnClient`-wrapper, feel free to! When the above PR has landed, we might want to note the exact version in these docs? I'm totally fine with merging these docs first and changing that later too though. :smile:

File: docs/guides/ssr.md
Comment (TheGooner93) [inline docs/guides/ssr.md:209]: @efilion is there a V3 version of this ? `query-core` is a V4 module ?

Comment (TkDodo) [reply docs/guides/ssr.md:209]: no. we're writing the docs for v4 here :)

Comment (TheGooner93) [reply docs/guides/ssr.md:209]: No worries, it made sense for me to just update to V4

**Comment by Ephem** [comment]
I took the liberty to push some changes to this, I hope that was okay.

* Removed `HydrateOnClient` in favour of `Hydrate` (that PR is ready to be merged now`
* Rewrote the Suspense-section
* Some minor tweaks, including removing the note about importing directly from the core package since I'm not sure that still applies after we merge the 'use client' PR
* Prettier
* `main` had new docs on integrating with Remix, so I moved this whole section down to below those (since this is beta stuff) and added a note to the Next-section

I think with those changes this should be ready to merge after https://github.com/TanStack/query/pull/4738, but I'd love it if someone could take a quick look. :smile: Great work on this @efilion!

**Review by Ephem — APPROVED** [review]

File: docs/react/guides/ssr.md
Comment (TkDodo) [inline docs/react/guides/ssr.md:428]: we should really import dehydrate from react-query, as it contains the whole query-core. Users won't have the query-core installed separately

**Review by TkDodo — APPROVED** [review]
Wow this is really good 👍

**Comment by efilion** [comment]
Suspense section is very well put. Thanks for taking care of that! Thrilled to see this PR get shipped :)

**Comment by hakankaan** [comment]
image
Hi. I'm trying to use Hydrate on server components but getting this error. Used HydrateOnClient component under QueryClient. Wrapped HydrateOnClient with another component and used at app/layout.tsx. I'm trying to achieve auth mechanism for routes. Can you guide me please?

image

**Comment by StringKe** [comment]
> image
> 
> Hi. I'm trying to use Hydrate on server components but getting this error. Used HydrateOnClient component under QueryClient. Wrapped HydrateOnClient with another component and used at app/layout.tsx. I'm trying to achieve auth mechanism for routes. Can you guide me please?
> image

I encountered the same problem, did you solve it?

**Comment by Ephem** [comment]
There's an issue to track this here, which includes a work around: https://github.com/TanStack/query/issues/4933

**Pull request merged**
